Upgrade Rails to 5.2

deivid-rodriguez opened the pull request:
#### :tophat: What? Why?

While working on some managed users features and refactorings, I started running once again into https://github.com/rspec/rspec-rails/issues/1887. It's a very annoying issue, and I was doing step by step refactoring and thus running the tests very often.. . It was unclear to me whether the fix was provided in Rails 5.2 or capybara 3.0, so I upgraded both :)

When pushing the branch with the upgrades, I run into some flakies, the classic "server timeout on the first test". I decided to investigate.

Those flakies usually mean that `sprockets-rails` is precompiling the assets on the fly in the first request and that took too long. There's something weird going on in there because I thought by precompiling the assets in the `build_test_app` job and sharing the precompiled assets with the other jobs we should be avoiding exactly that. But those flakies are there, so assuming they're due to asset precompilation (I've never seen other situation that affects only the first request of a test run), that means something is not happening as we expect.

To fix it, I chose to prevent that on the fly asset precompilation from happening. By doing that, we should fix the flakies but also:

* We no longer need the asset precompilation on the `build_test_app` job. Since that job is the most bottlenecking one of all, that means around 1min savings for every build!
* We no longer precompile assets on the development environment either. That means fast page loading on the first request of a freshly created development app! 

#### :pushpin: Related Issues
- Related to #3218.

#### :clipboard: Subtasks
- [x] Add `CHANGELOG` entry
- [ ] Another subtask

### :camera: Screenshots (optional)
_None_.

mrcasals commented on decidim-core/lib/tasks/decidim_tasks.rake:
```diff
@@ -2,5 +2,27 @@
+  task upgrade: [:choose_target_plugins, :"railties:install:migrations"]
+
+  desc "Setup environment so that only decidim migrations are installed."
+  task :choose_target_plugins do
```
I'm not sure about this change. If we're supporting Rails 5.2 now, shouldn't we allow the ActiveStorage migrations to be installed? 😕

josepjaume replied:
The original intent of `decidim:upgrade` was to `only` upgrade decidim. Maybe we could explicitly add the `ActiveStorage` ones (or write some instructions on the upgrade guide) once we make use of them, but I think it's fine the way it is now.

oriolgual replied:
Maybe I'm missing something, what this (specific) changes have to do with ActiveStorage?

deivid-rodriguez replied:
Hei! The thing is that this `activestorage` plugin is shipped as an engine with its own migrations. And this task was previously installing all migrations from all available engines (thus including `activestorage` ones).

I agree with @josepjaume's comment, but the real motivation for this change is not correctness or saving a couple of unused tables. Without this, I was not able to add new migrations to the test app, it would spit errors related to `activestorage` tables, not sure why...

deivid-rodriguez commented:
@josepjaume Good call, but I think I can extract the CI speed up to a separate PR since that's low risk and we can benefit from that right now. Sounds good?

mrcasals commented:
@deivid-rodriguez sounds good to me!

deivid-rodriguez commented:
I rebased this after all the new changes in master. I added a few things:

* I was getting a test hang on API tests, reproducible both locally and in CI. I added 
b8a8195 to fix it. What do you think @josepjaume? By the way, I found the `mail` gem doing [this trick](https://github.com/mikel/mail/blob/7ef5694b77bbe86be851dbcfae38b7b47c689555/lib/mail.rb#L40-L56) (see also https://github.com/rails/rails/pull/32808). Maybe it's cleaner than your approach to "keep track of orphans". Or maybe it's just not related, I thought it was worth mentioning though.

* I tried to fix a flaky on system specs in https://github.com/decidim/decidim/pull/3225/commits/62b7ea5d96c0d7a7c73c55e0d0ec3fab82bae145.

* Due to all the flakies I've seen lately, I added 6a8ee87 and 3e2fc17 to make investigating them easier. The first one seems to work fine, I've been using it for a while locally (see  https://github.com/rails/rails/pull/31960). The second one I'm not sure since I haven't got flakies since I added it.

deivid-rodriguez commented:
@mrcasals No problem! My main interest in this PR is that tests hang whenever I click Ctrl-C during a test run, and that's fixed in capybara 3.0. I found a way to upgrade to use capybara 3.0 without having to upgrade to 5.2 (it requires the test app to use the latest, unreleased, 5-1-stable branch from rails though). Is that acceptable for master at the moment?

So will you be generating an application from _master_ for that big demo? You naughty boys & girls :smile: ...

mrcasals commented:
@deivid-rodriguez  I've merged #3361! 😄

mrcasals commented:
@deivid-rodriguez this is not being blocked anymore! Do you need anything from us here? (reviews apart)

deivid-rodriguez commented:
Nice! I've been rebasing this branch once in a while for a while, and it's been consistently green. Let's see if it stays like that.

mrcasals commented:
@deivid-rodriguez should we merge then? 😄
